### PR TITLE
PyTorchLightningPruningCallback from Callback

### PR DESCRIPTION
**Pull request by TezRomacH**
## Motivation

PoC for #2315
Now PyTorchLightningPruningCallback is just a regular Callback instance

### CODE REVIEW

**Comment by toshihikoyanase** [comment]
@nzw0301 @HideakiImamura Could you review this PR since you reviewed #2294? If you don't have time enough, please feel free to remove the assignment.

**Comment by nzw0301** [comment]
Sorry, i'm busy this week...

File: examples/pytorch_lightning_simple.py [hunk 1]
```diff
@@ -81,58 +68,80 @@ def forward(self, data):
...
 
-    def forward(self, data):
+        # We optimize the number of layers, hidden units in each layer and dropouts.
+        n_layers = trial.suggest_int("n_layers", 1, 3)
+        dropout = trial.suggest_float("dropout", 0.2, 0.5)
+        output_dims = [
+            trial.suggest_int("n_units_l{}".format(i), 4, 128, log=True) for i in range(n_layers)
+        ]
+
+        self.model = Net(dropout, output_dims)
```
Comment (crcrpar) [inline examples/pytorch_lightning_simple.py:79]: I think we can define `self.model` using `torch.nn.Sequential`, which will be more concise.
Could you kindly do it?

File: examples/pytorch_lightning_simple.py [hunk 2]
```diff
@@ -81,58 +68,80 @@ def forward(self, data):
...
 
     trainer = pl.Trainer(
         logger=False,
         limit_val_batches=PERCENT_VALID_EXAMPLES,
-        checkpoint_callback=checkpoint_callback,
+        checkpoint_callback=False,
         max_epochs=EPOCHS,
-        gpus=1 if torch.cuda.is_available() else None,
-        callbacks=[PyTorchLightningPruningCallback(trial, monitor="val_acc", mode="max")],
+        gpus=-1 if torch.cuda.is_available() else None,
```
Comment (crcrpar) [inline examples/pytorch_lightning_simple.py:141]: I think it's better to let this program use a GPU if available. We can disable the use of GPUs outside the program by using `CUDA_VISIBLE_DEVICES`.
```suggestion
        gpus=1 if torch.cuda.is_available() else None,
```

**Review by crcrpar — COMMENTED** [review]
Thank you for the work.

It basically is cool 🙂

File: examples/pytorch_lightning_simple.py (hunk 2 above)
Comment (TezRomacH) [reply examples/pytorch_lightning_simple.py:141]: Setting `-1` in PL says that you use all available GPUs :)
https://pytorch-lightning.readthedocs.io/en/stable/trainer.html#gpus
Screenshot 2021-02-08 at 16 09 21

File: examples/pytorch_lightning_simple.py (hunk 1 above)
Comment (TezRomacH) [reply examples/pytorch_lightning_simple.py:79]: I thought for myself that it was better for the users to see that their model can be any `torch.nn.Module`. In Lightning we create model and experimenting with Trial.

Are you proposing to build a model inside Lightning?

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. It looks like there are a lot of changes being made that have nothing to do with the original purpose of the PR; I think one PR should do one thing, so can you revert the unrelated changes?

**Comment by TezRomacH** [comment]
> Thanks for the PR. It looks like there are a lot of changes being made that have nothing to do with the original purpose of the PR; I think one PR should do one thing, so can you revert the unrelated changes?

You're suggesting to separate updated example and updated callback into to two PRs?

**Comment by HideakiImamura** [comment]
Yes, I think we should separate them. Furthermore, I doubt that the changes being made in this PR on the example are necessary. For example, Optuna's example does not have type hinting, nor is it checked by CI.

File: examples/pytorch_lightning_simple.py (hunk 2 above)
Comment (crcrpar) [reply examples/pytorch_lightning_simple.py:141]: Oh, thank you. Obviously I missed that.

**Comment by HideakiImamura** [comment]
Sorry for the confusion. I had a discussion with a team member about type hinting in `examples`. It's OK that examples may or may not have type hinting. If you could make a separate PR for changes to examples and changes to integration, that would be much appreciated.

**Comment by TezRomacH** [comment]
Done :)

**Comment by TezRomacH** [comment]
Example is here https://github.com/optuna/optuna/pull/2332

**Review by HideakiImamura — APPROVED** [review]
Thanks for the simplification! LGTM. I'll review another PR.

**Review by crcrpar — APPROVED** [review]
looks great :)

**Pull request merged**